src/emoji/data.js

```diff
@@ -1,14 +1,70 @@
 /* @flow strict-local */
 import * as typeahead from '@zulip/shared/js/typeahead';
 
-import type { EmojiType, ReactionType, EmojiForShared } from '../types';
+import * as api from '../api';
+import * as logging from '../utils/logging';
+import type {
+  ThunkAction,
+  PerAccountAction,
+  ServerEmojiData,
+  EmojiType,
+  ReactionType,
+  EmojiForShared,
+} from '../types';
+import { tryParseUrl } from '../utils/url';
+import { REFRESH_SERVER_EMOJI_DATA } from '../actionConstants';
+import { NetworkError } from '../api/apiErrors';
 import { objectFromEntries } from '../jsBackport';
 import { unicodeCodeByName, override } from './codePointMap';
 import zulipExtraEmojiMap from './zulipExtraEmojiMap';
 import { objectEntries } from '../flowPonyfill';
 
 const unicodeEmojiNames = Object.keys(unicodeCodeByName);
 
+const refreshServerEmojiData = (data: ServerEmojiData): PerAccountAction => ({
+  type: REFRESH_SERVER_EMOJI_DATA,
+  data,
+});
+
+export const maybeRefreshServerEmojiData =
+  (server_emoji_data_url: string | void): ThunkAction<Promise<void>> =>
+  async (dispatch, getState) => {
+    if (server_emoji_data_url === undefined) {
+      // The server is too old to support this feature.
+      // TODO(server-6.0): Simplify away; should always be present.
+      return;
+    }
+    const parsedUrl = tryParseUrl(server_emoji_data_url);
+    if (!parsedUrl) {
+      logging.error('Invalid URL at server_emoji_data_url in /register response');
+      return;
+    }
+
+    let data = undefined;
+    try {
+      data = await api.fetchServerEmojiData(parsedUrl);
+    } catch (errorIllTyped) {
+      const e: mixed = errorIllTyped; // https://github.com/facebook/flow/issues/2470
+
+      if (!(e instanceof Error)) {
+        logging.error('Unexpected non-error thrown from api.fetchServerEmojiData.');
+        return;
+      }
+
+      if (e instanceof NetworkError) {
+        // Probably a user-connectivity error; no need to log. Hope it goes
+        // through next time.
+        // TODO(?): Could retry?
```

Comment:
Yeah, I think best to retry -- that avoids a bad experience if you're on a flaky connection when things first load.

Should be able to use the existing `tryFetch`, to avoid needing to write any tricky retry logic a second time.